sycl/include/CL/sycl/intel/atomic_accessor.hpp

```diff
@@ -0,0 +1,128 @@
+inline constexpr scope_tag_t<memory_scope::device> device_scope{};
+inline constexpr scope_tag_t<memory_scope::system> system_scope{};
+
+#endif
+
+template <typename DataT, int Dimensions, memory_order DefaultOrder,
+          memory_scope DefaultScope,
+          access::target AccessTarget = access::target::global_buffer,
+          access::placeholder IsPlaceholder = access::placeholder::false_t>
+class atomic_accessor
+    : public accessor<DataT, Dimensions, access::mode::read_write, AccessTarget,
+                      IsPlaceholder> {
+
+  using AccessorT = accessor<DataT, Dimensions, access::mode::read_write,
+                             AccessTarget, IsPlaceholder>;
+
+private:
+  using AccessorT::getLinearIndex;
+  using AccessorT::getQualifiedPtr;
+
+  // Prevent non-atomic access to atomic accessor
+  multi_ptr<DataT, AccessorT::AS> get_pointer() const = delete;
+
+protected:
+  using AccessorT::AdjustedDim;
+
+public:
+  using value_type = DataT;
+  using reference =
+      atomic_ref<DataT, DefaultOrder, DefaultScope, AccessorT::AS>;
+
+  using AccessorT::AccessorT;
+
+#if __cplusplus > 201402L
+
+  template <typename T = DataT, int Dims = Dimensions, typename AllocatorT,
+            memory_order Order, memory_scope Scope>
+  atomic_accessor(buffer<T, Dims, AllocatorT> &BufferRef, order_tag_t<Order>,
+                  scope_tag_t<Scope>, const property_list &PropertyList = {})
+      : atomic_accessor(BufferRef, PropertyList) {}
+
+  template <typename T = DataT, int Dims = Dimensions, typename AllocatorT,
+            memory_order Order, memory_scope Scope>
+  atomic_accessor(buffer<T, Dims, AllocatorT> &BufferRef,
+                  handler &CommandGroupHandler, order_tag_t<Order>,
+                  scope_tag_t<Scope>, const property_list &PropertyList = {})
+      : atomic_accessor(BufferRef, CommandGroupHandler, PropertyList) {}
+
+#endif
+
+  // Override subscript operators and conversions to wrap in an atomic_ref
+  template <int Dims = Dimensions>
+  operator typename detail::enable_if_t<Dims == 0, reference>() const {
+    const size_t LinearIndex = getLinearIndex(id<AdjustedDim>());
+    return reference(getQualifiedPtr()[LinearIndex]);
+  }
+
+  template <int Dims = Dimensions>
+  typename detail::enable_if_t<(Dims > 0), reference>
+  operator[](id<Dimensions> Index) const {
+    const size_t LinearIndex = getLinearIndex(Index);
+    return reference(getQualifiedPtr()[LinearIndex]);
+  }
+
+  template <int Dims = Dimensions>
+  typename detail::enable_if_t<Dims == 1, reference>
+  operator[](size_t Index) const {
+    const size_t LinearIndex = getLinearIndex(id<AdjustedDim>(Index));
+    return reference(getQualifiedPtr()[LinearIndex]);
+  }
+};
+
+#if __cplusplus > 201402L
+
+template <typename DataT, int Dimensions, typename AllocatorT,
+          memory_order Order, memory_scope Scope, typename... Ts>
+atomic_accessor(buffer<DataT, Dimensions, AllocatorT>, Ts...,
```

Comment:
Why do you have `Ts...` here - you have only 2 ctors in place in the class.

I found this approach non working - when deduction is applied all Ts... are always considered zero length.
Since it is not needed here, its probably better be removed, to not getting into this trap later.

I applied the same approach initially, but moved to uglier approach in the later implementation.

Other than that - LGTM

Reply:
Thanks, fixed in https://github.com/intel/llvm/pull/2069/commits/946192e1d6682d1831646d0ee1c8b8765df5202e.